TFEngine.py

```diff
@@ -1857,86 +1857,127 @@ def search(self, dataset, do_eval=True, output_layer_name="output", output_file=
+
+      In addition, for each output layer, we expect the following parameters in kwargs:
+        list[numpy.ndarray] output_<layer name>
+        list[numpy.ndarray] beam_scores_<layer name>
+        list[numpy.ndarray] target_<target key>
       """
+
+      outputs, beam_scores, targets = [], [], []
+      for t in range(num_targets):
+        outputs.append(kwargs["output_" + output_layer_names[t]])
+        beam_scores.append(kwargs["beam_scores_" + output_layer_names[t]])
+        targets.append(kwargs["target_" + target_keys[t]])
+
       n_batch = len(seq_idx)  # without beam
       assert n_batch == len(seq_tag)
-      assert n_batch * (out_beam_size or 1) == len(output)
-      if targets is not None:
-        assert n_batch == len(targets)
-      if beam_scores is not None:
-        assert beam_scores.shape == (n_batch, out_beam_size)
-      if output_layer.output.dim == 256 and output_layer.output.sparse:
-        # Interpret output as bytes/utf8-string.
-        output = [bytearray(o).decode("utf8") for o in output]
+
+      for t in range(num_targets):
+        if beam_scores[t] is not None:
+          assert beam_scores[t].shape == (n_batch, out_beam_sizes[t])
+
+        assert n_batch * (out_beam_sizes[t] or 1) == len(outputs[t])
+        if targets[t] is not None:
+          assert n_batch == len(targets[t])
+
+        if output_layers[t].output.dim == 256 and output_layers[t].output.sparse:
+          # Interpret output as bytes/utf8-string.
+          outputs[t] = bytearray(outputs[t]).decode("utf8")
+
       for i in range(len(seq_idx)):
-        if out_beam_size is None:
-          print("seq_idx: %i, seq_tag: %r, output: %r" % (seq_idx[i], seq_tag[i], output[i]), file=log.v4)
-          out_idx = i
-        else:
-          print("seq_idx: %i, seq_tag: %r, outputs: %r" % (
-            seq_idx[i], seq_tag[i], output[i * out_beam_size:(i + 1)*out_beam_size]), file=log.v4)
-          out_idx = i * out_beam_size
-        if target_key and dataset.can_serialize_data(target_key):
-          print("  ref:", dataset.serialize_data(key=target_key, data=targets[i]), file=log.v4)
-          if out_beam_size is None:
-            print("  hyp:", dataset.serialize_data(key=target_key, data=output[out_idx]), file=log.v4)
-          else:
-            assert beam_scores is not None
-            for b in range(out_beam_size):
-              print(
-                "  hyp %i, score %f:" % (b, beam_scores[i][b]),
-                dataset.serialize_data(key=target_key, data=output[out_idx + b]),
-                file=log.v4)
         if out_cache is not None:
           corpus_seq_idx = dataset.get_corpus_seq_idx(seq_idx[i])
           assert corpus_seq_idx not in out_cache
           seq_idx_to_tag[corpus_seq_idx] = seq_tag[i]
-          if out_beam_size is None:
-            out_cache[corpus_seq_idx] = dataset.serialize_data(key=target_key, data=output[out_idx])
+          if num_serializable_targets > 1:
```

Comment:
The check should be `isinstance(output_layer_names, dict)`, and not `num_serializable_targets == 1`.